packages/react-core/src/components/Menu/MenuItem.tsx

```diff
@@ -77,20 +84,85 @@ const MenuItemBase: React.FunctionComponent<MenuItemProps> = ({
   icon,
   actions,
   onShowFlyout,
-  innerRef,
   drilldownMenu,
   isOnPath,
   ...props
 }: MenuItemProps) => {
+  const {
+    menuId,
+    parentMenu,
+    onSelect,
+    onActionClick,
+    activeItemId,
+    selected,
+    drilldownItemPath,
+    onDrillIn,
+    onDrillOut,
+    flyoutRef,
+    setFlyoutRef
+  } = React.useContext(MenuContext);
   const Component = component || to ? 'a' : 'button';
-  const [flyoutVisible, setFlyoutVisible] = React.useState(false);
   const [flyoutTarget, setFlyoutTarget] = React.useState(null);
+  const flyoutContext = React.useContext(FlyoutContext);
+  const [flyoutXDirection, setFlyoutXDirection] = React.useState(flyoutContext.direction);
+  const ref = React.useRef<HTMLLIElement>();
+  const flyoutVisible = ref === flyoutRef;
 
-  const showFlyout = (displayFlyout: boolean) => {
-    setFlyoutVisible(displayFlyout);
-    onShowFlyout && displayFlyout && onShowFlyout();
+  const hasFlyout = flyoutMenu !== undefined;
+  const showFlyout = (show: boolean) => {
+    if (!flyoutVisible && show) {
+      setFlyoutRef(ref);
+    } else if (flyoutVisible && !show) {
+      setFlyoutRef(null);
+    }
+    onShowFlyout && show && onShowFlyout();
   };
 
+  useIsomorphicLayoutEffect(() => {
+    if (hasFlyout && ref.current && canUseDOM) {
+      const flyoutMenu = ref.current.lastElementChild as HTMLElement;
+      if (flyoutMenu && flyoutMenu.classList.contains(styles.menu)) {
+        const origin = ref.current.getClientRects()[0];
+        const rect = flyoutMenu.getClientRects()[0];
+        if (origin && rect) {
+          const spaceLeftLeft = origin.x - rect.width;
+          const spaceLeftRight = window.innerWidth - origin.x - origin.width - rect.width;
+          let xDir = flyoutXDirection as 'left' | 'right' | 'none';
+          if (spaceLeftRight < 0 && xDir !== 'left') {
+            setFlyoutXDirection('left');
+            xDir = 'left';
+          } else if (spaceLeftLeft < 0 && xDir !== 'right') {
+            setFlyoutXDirection('right');
+            xDir = 'right';
+          }
+          let xOffset = 0;
+          if (spaceLeftLeft < 0 && spaceLeftRight < 0) {
+            xOffset = xDir === 'right' ? -spaceLeftRight : -spaceLeftLeft;
+          }
+          if (xDir === 'left') {
+            flyoutMenu.classList.add(styles.modifiers.left);
+            flyoutMenu.style.setProperty(rightOffset.name, `-${xOffset}px`);
+          } else {
+            flyoutMenu.style.setProperty(leftOffset.name, `-${xOffset}px`);
+          }
+
+          const spaceLeftBot = window.innerHeight - origin.y - rect.height;
+          const spaceLeftTop = window.innerHeight - rect.height;
+          if (spaceLeftTop < 0 && spaceLeftBot < 0) {
+            // working idea: page can usually scroll down, but not up
+            // TODO: proper scroll buttons
```

Comment:
Is this going to be implemented now or in a follow up?

Reply:
There are no designs or HTML/CSS for vertical menu scroll buttons so I'd say in a followup if a future issue is opened.